File: cmd/ocdev/tests/test_list_json.py
#!/usr/bin/env python3
"""Fake-Incus CLI tests: python3 cmd/ocdev/tests/test_list_json.py ./bin/ocdev.

Build the repository binary first. No real Incus executable is reachable.
"""
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest

BINARY = Path(sys.argv.pop(1)).resolve() if len(sys.argv) > 1 else None


class ListJsonTests(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp.cleanup)
        self.root = Path(self.temp.name)
        self.home = self.root / "home"
        self.home.mkdir()
        self.bin = self.root / "bin"
        self.bin.mkdir()
        self.env = dict(os.environ, HOME=str(self.home), PATH=str(self.bin),
                        FAKE_ROOT=str(self.root))
        self.script("groups", "print('incus-admin')\n")
        self.script("incus", '''import json, os, pathlib, sys
root = pathlib.Path(os.environ["FAKE_ROOT"])
with (root / "calls").open("a") as log:
    log.write(json.dumps(sys.argv[1:]) + "\\n")
if sys.argv[1:] not in (["list", "--format=json", "ocdev-"],
                        ["list", "--format=csv", "-c", "n,s", "ocdev-"]):
    print("Unexpected Incus command", file=sys.stderr)
    sys.exit(99)
sys.stdout.write((root / "response").read_text())
sys.stderr.write((root / "stderr").read_text())
sys.exit(int((root / "exit").read_text()))
''')
        self.response([])

    def script(self, name, body):
        path = self.bin / name
        path.write_text(f"#!{sys.executable}\n" + body)
        path.chmod(0o755)

    def response(self, data, code=0, stderr="", raw=False):
        (self.root / "response").write_text(data if raw else json.dumps(data))
        (self.root / "exit").write_text(str(code))
        (self.root / "stderr").write_text(stderr)

    def run_list(self, *args):
        return subprocess.run([str(BINARY), "list", *args], env=self.env,
                              text=True, capture_output=True, timeout=10)

    def assert_failure(self, result):
        self.assertNotEqual(result.returncode, 0)
        self.assertEqual(result.stdout, "")
        self.assertTrue(result.stderr)
        self.assertNotIn("private-sentinel", result.stderr)

    def test_projection_and_prefix_and_ports(self):
        state = self.home / ".ocdev"
        state.mkdir()
        ports = state / "ports"
        ports.write_text("demo-ocdev-copy:2210\nstopped:2220\n")
        before = ports.read_bytes()
        self.response([
            {"name": "ocdev-demo-ocdev-copy", "status": "Running",
             "config": {"volatile.uuid": "dummy-uuid", "user.private": "private-sentinel"},
             "expanded_config": {"user.private": "private-sentinel"}},
            {"name": "ocdev-stopped", "status": "Stopped", "config": {}},
            {"name": "other-ocdev-demo", "status": "Running"},
        ])
        result = self.run_list("--json")
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(result.stderr, "")
        self.assertEqual(json.loads(result.stdout), [
            dict(name="demo-ocdev-copy", instance="ocdev-demo-ocdev-copy",
                 status="Running", uuid="dummy-uuid", ssh_port=2210),
            dict(name="stopped", instance="ocdev-stopped", status="Stopped",
                 uuid=None, ssh_port=2220),
        ])
        self.assertEqual(ports.read_bytes(), before)
        self.assertEqual(sorted(p.name for p in state.iterdir()), ["ports"])
        self.assertEqual((self.root / "calls").read_text(),
                         '["list", "--format=json", "ocdev-"]\n')

    def test_missing_uuid_and_allocation_no_state_created(self):
        for config in (None, {}, {"volatile.uuid": ""}, {"volatile.uuid": None}):
            with self.subTest(config=config):
                self.response([dict(name="ocdev-demo", status="Frozen", config=config)])
                result = self.run_list("--json")
                self.assertEqual(result.returncode, 0, result.stderr)
                self.assertEqual(json.loads(result.stdout), [dict(
                    name="demo", instance="ocdev-demo", status="Frozen",
                    uuid=None, ssh_port=None)])
                self.assertFalse((self.home / ".ocdev").exists())
        self.response([dict(name="ocdev-demo", status="Running")])
        self.assertIsNone(json.loads(self.run_list("--json").stdout)[0]["uuid"])

    def test_unavailable_allocations_are_null(self):
        state = self.home / ".ocdev"
        state.mkdir()
        self.response([dict(name="ocdev-demo", status="Running")])
        for allocation in ("other:2200\n", "demo:invalid\n", "demo:0\n", "demo:-1\n"):
            with self.subTest(allocation=allocation):
                (state / "ports").write_text(allocation)
                result = self.run_list("--json")
                self.assertEqual(result.returncode, 0, result.stderr)
                self.assertIsNone(json.loads(result.stdout)[0]["ssh_port"])

    def test_empty_success(self):
        result = self.run_list("--json")
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(result.stdout, "[]\n")
        self.assertEqual(result.stderr, "")
        self.assertFalse((self.home / ".ocdev").exists())

    def test_nonmatching_instances_are_empty(self):
        self.response([dict(name="unrelated", status="Running")])
        self.assertEqual(self.run_list("--json").stdout, "[]\n")

    def test_query_failure_is_not_empty_success(self):
        for output in ("[]", "No container found"):
            self.response(output, code=1, stderr="query failed private-sentinel\n", raw=True)
            self.assert_failure(self.run_list("--json"))

    def test_warning_does_not_corrupt_json(self):
        self.response([], stderr="warning from Incus\n")
        result = self.run_list("--json")
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(result.stdout, "[]\n")
        self.assertEqual(result.stderr, "")

    def test_malformed_metadata_never_emits_partial_json(self):
        valid = dict(name="ocdev-good", status="Running")
        for bad in ("", "not json private-sentinel", "{}", "null", '[42]',
                    json.dumps([valid, dict(name="ocdev-bad", status=12)]),
                    json.dumps([dict(status="Running")]),
                    json.dumps([dict(name="ocdev-demo", status="Running", config=[])]),
                    json.dumps([dict(name="ocdev-demo", status="Running",
                                     config={"volatile.uuid": 12})])):
            with self.subTest(data=bad):
                self.response(bad, raw=True)
                self.assert_failure(self.run_list("--json"))

    def test_prerequisite_failures(self):
        self.script("groups", "print('users')\n")
        self.assert_failure(self.run_list("--json"))
        self.assertFalse((self.root / "calls").exists())
        (self.bin / "incus").unlink()
        self.assert_failure(self.run_list("--json"))
        self.assertFalse((self.home / ".ocdev").exists())

    def test_default_table_unchanged(self):
        self.response("ocdev-demo,RUNNING\n", raw=True)
        result = self.run_list()
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(result.stdout,
                         f"{'NAME':<20} {'STATUS':<10} SSH PORT\n"
                         f"{'demo':<20} {'RUNNING':<10} N/A\n")
        self.assertEqual((self.root / "calls").read_text(),
                         '["list", "--format=csv", "-c", "n,s", "ocdev-"]\n')


if __name__ == "__main__":
    if BINARY is None or not BINARY.is_file():
        sys.exit("Pass a freshly built repository binary; no installed binary is used.")
    unittest.main()

File: cmd/ocdev/tests/test_list_json.nim
## Black-box JSON listing contracts. Only the supplied compiled fake is on PATH.
import std/[os, json, unittest, strutils, algorithm]
import test_support

# Binary arguments are fixture inputs, not unittest name filters.
disableParamFiltering()
let args = commandLineParams()
if args.len != 2:
  quit("Usage: test-list-json <built ocdev binary> <built fake Incus>", 2)
let binary = absolutePath(args[0])
let fake = absolutePath(args[1])

proc responseRaw(s: Sandbox; data: string; code = 0; error = "") =
  writeFile(s.home / "response", data)
  writeFile(s.home / "exit", $code)
  writeFile(s.home / "stderr", error)

proc response(s: Sandbox; data: JsonNode; code = 0; error = "") =
  responseRaw(s, $data, code, error)

proc runList(s: Sandbox; args: seq[string] = @[]): RunResult =
  run(s, binary, @["list"] & args, timeoutMs = 10000)

proc assertFailure(r: RunResult) =
  check r.code != 0
  check r.output == ""
  check r.error.len > 0
  check "private-sentinel" notin r.error

proc calls(s: Sandbox): seq[JsonNode] =
  for line in readFile(s.home / "calls").splitLines():
    if line.len > 0: result.add(parseJson(line))

suite "list JSON with compiled fake Incus":
  setup:
    let s = newSandbox("ocdev-list-", fake)
    s.env["FAKE_ROOT"] = s.home
    s.env["PATH"] = s.home
    copyFile(s.home / "incus", s.home / "groups")
    setFilePermissions(s.home / "groups", {fpUserRead, fpUserWrite, fpUserExec})
    response(s, newJArray())
  teardown:
    close(s)

  test "projection and prefix and ports":
    let state = s.home / ".ocdev"
    createDir(state)
    let ports = state / "ports"
    writeFile(ports, "demo-ocdev-copy:2210\nstopped:2220\n")
    let before = readFile(ports)
    response(s, %*[
      {"name": "ocdev-demo-ocdev-copy", "status": "Running",
       "config": {"volatile.uuid": "dummy-uuid", "user.private": "private-sentinel"},
       "expanded_config": {"user.private": "private-sentinel"}},
      {"name": "ocdev-stopped", "status": "Stopped", "config": {}},
      {"name": "other-ocdev-demo", "status": "Running"}])
    let r = runList(s, @["--json"])
    checkpoint r.error
    check r.code == 0
    check r.error == ""
    check parseJson(r.output) == %*[
      {"name": "demo-ocdev-copy", "instance": "ocdev-demo-ocdev-copy",
       "status": "Running", "uuid": "dummy-uuid", "ssh_port": 2210},
      {"name": "stopped", "instance": "ocdev-stopped", "status": "Stopped",
       "uuid": nil, "ssh_port": 2220}]
    check readFile(ports) == before
    var children: seq[string]
    for kind, path in walkDir(state): children.add(extractFilename(path))
    children.sort()
    check children == @["ports"]
    check calls(s) == @[%*["list", "--format=json", "ocdev-"]]

  test "missing UUID and allocation never initialize state":
    for config in [newJNull(), newJObject(), %*{"volatile.uuid": ""},
                   %*{"volatile.uuid": nil}]:
      response(s, %*[{"name": "ocdev-demo", "status": "Frozen", "config": config}])
      let r = runList(s, @["--json"])
      checkpoint r.error
      check r.code == 0
      check parseJson(r.output) == %*[{"name": "demo", "instance": "ocdev-demo",
        "status": "Frozen", "uuid": nil, "ssh_port": nil}]
      check not pathExists(s.home / ".ocdev")
    response(s, %*[{"name": "ocdev-demo", "status": "Running"}])
    check parseJson(runList(s, @["--json"]).output)[0]["uuid"].kind == JNull

  test "unavailable allocations are null":
    let state = s.home / ".ocdev"
    createDir(state)
    response(s, %*[{"name": "ocdev-demo", "status": "Running"}])
    for allocation in ["other:2200\n", "demo:invalid\n", "demo:0\n", "demo:-1\n"]:
      writeFile(state / "ports", allocation)
      let r = runList(s, @["--json"])
      checkpoint r.error
      check r.code == 0
      check parseJson(r.output)[0]["ssh_port"].kind == JNull

  test "empty success":
    let r = runList(s, @["--json"])
    checkpoint r.error
    check r.code == 0
    check r.output == "[]\n"
    check r.error == ""
    check not pathExists(s.home / ".ocdev")

  test "nonmatching instances are empty":
    response(s, %*[{"name": "unrelated", "status": "Running"}])
    check runList(s, @["--json"]).output == "[]\n"

  test "query failure is not empty success":
    for output in ["[]", "No container found"]:
      responseRaw(s, output, code = 1, error = "query failed private-sentinel\n")
      assertFailure(runList(s, @["--json"]))

  test "warning does not corrupt JSON":
    response(s, newJArray(), error = "warning from Incus\n")
    let r = runList(s, @["--json"])
    checkpoint r.error
    check r.code == 0
    check r.output == "[]\n"
    check r.error == ""

  test "malformed metadata never emits partial JSON":
    for bad in ["", "not json private-sentinel", "{}", "null", "[42]",
      $(%*[{"name": "ocdev-good", "status": "Running"}, {"name": "ocdev-bad", "status": 12}]),
      $(%*[{"status": "Running"}]),
      $(%*[{"name": "ocdev-demo", "status": "Running", "config": []}]),
      $(%*[{"name": "ocdev-demo", "status": "Running", "config": {"volatile.uuid": 12}}])]:
      responseRaw(s, bad)
      assertFailure(runList(s, @["--json"]))

  test "prerequisite failures":
    writeFile(s.home / "groups-output", "users")
    assertFailure(runList(s, @["--json"]))
    check not pathExists(s.home / "calls")
    removeFile(s.home / "incus")
    assertFailure(runList(s, @["--json"]))
    check not pathExists(s.home / ".ocdev")

  test "default table unchanged":
    responseRaw(s, "ocdev-demo,RUNNING\n")
    let r = runList(s)
    checkpoint r.error
    check r.code == 0
    check r.error == ""
    check r.output == "NAME".alignLeft(20) & " " & "STATUS".alignLeft(10) & " SSH PORT\n" &
      "demo".alignLeft(20) & " " & "RUNNING".alignLeft(10) & " N/A\n"
    check calls(s) == @[%*["list", "--format=csv", "-c", "n,s", "ocdev-"]]
